pil-analyzer/tests/condenser.rs

```diff
@@ -174,3 +174,95 @@ pub fn constructed_constraints() {
 "#;
     assert_eq!(formatted, expected);
 }
+
+#[test]
+#[should_panic = "There are pre-existing constraints or witness columns"]
+pub fn capture_stage_non_fresh_wit() {
+    let input = r#"
+        namespace std::prover;
+            let capture_stage: (-> int) -> Constr[] = 9;
+
+        namespace Main(1024);
+            col witness x;
+            let f = constr || { x = 2; 1024 };
+            std::prover::capture_stage(f);
+    "#;
+    analyze_string::<GoldilocksField>(input);
+}
+
+#[test]
+#[should_panic = "There are pre-existing constraints or witness columns"]
+pub fn capture_stage_non_fresh_inter() {
+    let input = r#"
+        namespace std::prover;
+            let capture_stage: (-> int) -> Constr[] = 9;
+
+        namespace Main(1024);
+            col x = 3;
+            let f = constr || { x = 2; 1024 };
+            std::prover::capture_stage(f);
+    "#;
+    analyze_string::<GoldilocksField>(input);
+}
+
+#[test]
+#[should_panic = "There are pre-existing constraints or witness columns"]
+pub fn capture_stage_non_fresh_constr() {
+    let input = r#"
+        namespace std::prover;
+            let capture_stage: (-> int) -> Constr[] = 9;
+
+        namespace Main(1024);
+            4 = 5;
+            let f = constr || 1024;
+            std::prover::capture_stage(f);
+    "#;
+    analyze_string::<GoldilocksField>(input);
+}
+
+#[test]
+#[should_panic = "returned degree 2048, but the degree has already been set to 1024."]
+pub fn capture_stage_set_different_degree() {
+    let input = r#"
+        namespace std::prover;
+            let capture_stage: (-> int) -> Constr[] = 9;
+
+        namespace Main(1024);
+            let f = constr || 2048;
+            std::prover::capture_stage(f);
+    "#;
+    analyze_string::<GoldilocksField>(input);
+}
+
+#[test]
+pub fn capture_stage_working() {
+    let input = r#"
+        namespace std::prover;
+            let capture_stage: (-> int) -> Constr[] = 9;
+
+        namespace Main;
+            let f = constr || { let x; let y; x + y = 2; x = 8; 1024 };
+            // Just use the second constraint.
+            std::prover::capture_stage(f)[1];
+            let w;
+            w = 10;
```

Comment:
I'd expect this to fail since there are other columns and constraints besides the call to `capture_stage`. Why doesn't it fail?

Reply:
I agree that this is a bit weird. In the end, you want to process the constraints that are returned by the call to `capture_stage`. Processing the constraints involves adding new constraints and witness column (but at a higher stage).

My current way to deal with it is that it's not OK to have witness columns or constraints above the call to `capture_stage`, but it is OK to have them below. The reason is that we only know the stage of witness columns once we return from `capture_stage`. Another way would be to retroactively change the stage of all witness columns that are already present in the condenser when we call `capture_stage`. This might indeed be a bit more uniform.

Reply:
Hmm yea this "above" and "below" separation is kinda weird. Why not have a function that receives the results of `capture_stage` as argument and then adds constraints on top of that?

Reply:
You mean like
```
capture_stage(constraint_generator, constraint_processor);
```
and `capture_stage` itself either returns nothing or it returns the processed constraints and it is illegal to have other witness column declarations or constraints at the global level if there is a call to `capture_stage`?

Reply:
yea, but not sure if that's over-thinking